src/polymer/polymer2-element-scanner.ts

```diff
@@ -65,161 +57,183 @@ export class Polymer2ElementScanner implements JavaScriptScanner {
   async scan(
       document: JavaScriptDocument,
       visit: (visitor: Visitor) => Promise<void>): Promise<ScannedElement[]> {
-    const visitor = new ElementVisitor(document);
-    await visit(visitor);
-    return visitor.getRegisteredElements();
-  }
-}
+    const classFinder = new ClassFinder();
+    const elementDefinitionFinder =
+        new CustomElementsDefineCallFinder(document);
+    // Find all classes and all calls to customElements.define()
+    await Promise.all([visit(classFinder), visit(elementDefinitionFinder)]);
+
+    const elementDefinitionsByClassName = new Map<string, ElementDefineCall>();
+    const elementDefinitionsByClassExpression =
+        new Map<estree.ClassExpression, ElementDefineCall>();
+
+    for (const defineCall of elementDefinitionFinder.calls) {
+      if (defineCall.clazz.type === 'MaybeChainedIdentifier') {
+        elementDefinitionsByClassName.set(defineCall.clazz.name, defineCall);
+      } else {
+        elementDefinitionsByClassExpression.set(defineCall.clazz, defineCall);
+      }
+    }
+    // TODO(rictic): emit ElementDefineCallFeatures for define calls that don't
+    //     map to any local classes?
+
+    const customElements: CustomElementDefinition[] = [];
+    const normalClasses = [];
+    for (const clazz of classFinder.classes) {
+      if (clazz.ast.type === 'ClassExpression') {
```

Comment:
There seems to be 3 separate cases here, it'd be nice to outline them with comments